File: python/src/config_builder.py
#!/usr/bin/env python

import yaml
import os
import logging
from .os_vars import OSVars
from .env_config import EnvConfig
from .env_conf_loader_factory import EnvConfigLoaderFactory
from .secrets import Secrets
from .logger import Logger

yaml_type_to_python = {"String": str, "Bool": bool, "Int": int, "Float": float}


class ConfigBuilder:
    def __init__(self):
        pass

    def __build_os_vars(self, data):
        for (env_var_name, env_var_data) in data["env-vars"].items():
            # print(f"var name: {env_var_name}, var data: {env_var_data}")
            if "is_mandatory" in env_var_data and env_var_data["is_mandatory"] is True:
                OSVars.register_mandatory(
                    env_var_name,
                    env_var_data["description"],
                    yaml_type_to_python[env_var_data["type"]],
                )
            else:
                default_val = None
                if "default" in env_var_data:
                    default_val = env_var_data["default"]

                OSVars.register(
                    env_var_name,
                    env_var_data["description"],
                    yaml_type_to_python[env_var_data["type"]],
                    default_val,
                )

    def __build_conf(self, data):
        if "config" not in data:
            return

        conf_data = data["config"]
        conf_provider_name = None
        if "provider" in conf_data:
            conf_provider_name = conf_data["provider"]

        conf_loader = EnvConfigLoaderFactory().get_loader(conf_provider_name)

        if "parent_environments" in conf_data:
            EnvConfig.instance().set_env_fallback(conf_data["parent_environments"])

        EnvConfig.instance().set_loader(conf_loader)

        if "categories" not in conf_data:
            return

        for category in conf_data["categories"]:
            EnvConfig.instance().require_category(category)

    def __build_secrets(self, data):
        if "secrets" not in data:
            return

        secrets_conf = data["secrets"]

        if "required" in secrets_conf:
            for secret_key in secrets_conf["required"]:
                try:
                    Secrets.get(secret_key)
                except Exception as e:
                    raise Exception(
                        f"Failed fetching Secrets key {secret_key}. Error: {e}"
                    )

    def __build_logger(self, data):
        logging.getLogger("requests").setLevel(logging.WARNING)

        log_level = None
        colored = False

        if "logger" in data:
            if "level" in data["logger"]:
                log_level = data["logger"]["level"]
            if "colored" in data["logger"]:
                colored = data["logger"]["colored"]

        Logger.instance().initialize(log_level, colored)

    def build(self, path_to_env_yaml=None):
        path = path_to_env_yaml
        if path_to_env_yaml is None:
            path = os.getcwd() + "/.envConfig.yml"

        print(f"Attempting to read env config yaml from {path}")
        env_file = open(path, "r")
        data = yaml.load(env_file, Loader=yaml.CLoader)
        # print(f"yaml data is {data}")

        self.__build_logger(data)

        self.__build_os_vars(data)
        OSVars.initialize()

        self.__build_secrets(data)

        self.__build_conf(data)

File: ruby/src/config_builder.rb
require 'yaml'
require_relative 'utils/logger'
require_relative 'os_vars'
require_relative 'env_config_loader_factory'
require_relative 'env_config'
require_relative 'secrets'

module TwistConf
  #############################################################################
  # GLOBALS and CONSTANTS                                                     #
  #############################################################################

  YAML_TYPE_TO_RUBY = {
    'String' => String,
    'Bool' => TrueClass,
    'Int' => Integer,
    'Float' => Float
  }

  #############################################################################
  # IMPLEMENTATION                                                            #
  #############################################################################

  class ConfigBuilder
    def __build_os_vars(data)
      if data['env-vars'].nil?
        return
      end

      data['env-vars'].each do |env_var_name, env_var_data|
        # puts "var name: #{env_var_name}, var data: #{env_var_data}"
        if env_var_data['is_mandatory'] == true
          OSVars.register_mandatory(
            env_var_name,
            env_var_data['description'],
            YAML_TYPE_TO_RUBY[env_var_data['type']]
          )
        else
          default_val = env_var_data['default']

          # registration expects string value (for when it comes from the real env as opposed to yaml)
          unless default_val.nil?
            default_val = default_val.to_s
          end

          OSVars.register(env_var_name, env_var_data['description'], YAML_TYPE_TO_RUBY[env_var_data['type']], default_val)
        end
      end
    end

    def __build_conf(data)
      if data['config'].nil?
        return
      end

      conf_data = data['config']
      conf_provider_name = conf_data['provider']
      factory = EnvConfigLoaderFactory.new
      conf_loader = factory.get_loader(conf_provider_name)

      if conf_data['parent_environments']
        EnvConfig.instance.set_env_fallback(conf_data['parent_environments'])
      end

      EnvConfig.instance.inject_loader(conf_loader)

      if conf_data['categories'].nil?
        return
      end

      conf_data['categories'].each do |category|
        EnvConfig.instance.require_category(category)
      end
    end

    def __build_secrets(data)
      if data['secrets'].nil?
        return
      end

      secrets_conf = data['secrets']

      return unless secrets_conf['required']

      secrets_conf['required'].each do |secret_key|
        # rubocop:disable Style/RedundantBegin
        begin
          Secrets.get(secret_key)
        rescue StandardError => e
          raise "Failed fetching Secrets key #{secret_key}. Error: #{e}"
        end
        # rubocop:enable Style/RedundantBegin
      end
    end

    def __build_logger(data)
      log_level = nil

      if data['logger']
        if data['logger']['level']
          log_level = data['logger']['level']
        end
        # not supported at the moment
        # if data['logger']['colored']
        #   colored = data['logger']['colored']
        # end
      end

      Log.instance.init(log_level)
    end

    def build(path_to_env_yaml = nil)
      actual_path = path_to_env_yaml
      if path_to_env_yaml.nil?
        actual_path = Dir.getwd + '/.envConfig.yml'
      end

      puts "Attempting to read env config yaml from #{actual_path}"
      env_file = File.open(actual_path)
      file_data = env_file.read
      env_file.close
      data = YAML.safe_load(file_data)

      # puts "yaml data is #{data}"

      __build_logger(data)

      __build_os_vars(data)
      OSVars.instance.init

      __build_secrets(data)

      __build_conf(data)
    end
  end
end
